slight/src/commands/run.rs

```diff
@@ -111,82 +134,163 @@ async fn shutdown_signal() {
 fn build_store_instance(
     toml: &TomlFile,
     toml_file_path: impl AsRef<Path>,
+    linked_capabilities: &mut HashSet<String>,
+    capability_store: &mut HashMap<String, BasicState>,
     resource_map: Arc<Mutex<StateTable>>,
     module: impl AsRef<Path>,
 ) -> Result<Builder> {
     let mut builder = Builder::new_default(module)?;
     let mut slight_builder = SlightCtxBuilder::default();
     builder.link_wasi()?;
-    if toml.specversion.as_ref().unwrap() == "0.1" {
+    if toml.specversion.as_ref().unwrap() == "0.2" {
         for c in toml.capability.as_ref().unwrap() {
-            let resource_type: &str = c.name.as_str();
+            let resource_type: &str = c.resource.as_str();
             match resource_type {
                 "events" => {
-                    builder.link_capability::<Events<Builder>>(resource_type.to_string())?;
-                    slight_builder =
-                        slight_builder.add_state(State::Events(EventsState::<Builder>::new(
-                            resource_map.clone(),
-                        )))?;
+                    if !linked_capabilities.contains("events") {
+                        builder.link_capability::<Events<Builder>>(resource_type.to_string())?;
+                        linked_capabilities.insert("events".to_string());
+
+                        slight_builder =
+                            slight_builder.add_state(State::Events(
+                                EventsState::<Builder>::new(resource_map.clone()),
+                            ))?;
+                    } else {
+                        bail!("the events capability was already linked");
+                    }
                 }
                 _ if KV_HOST_IMPLEMENTORS.contains(&resource_type) => {
-                    if let Some(ss) = &toml.secret_store {
+                    if !linked_capabilities.contains("kv") {
                         builder.link_capability::<Kv>("kv".to_string())?;
-                        slight_builder =
-                            slight_builder.add_state(State::Kv(slight_kv::KvState::new(
-                                resource_type.to_string(),
-                                BasicState::new(resource_map.clone(), ss, &toml_file_path),
-                            )))?;
+                        linked_capabilities.insert("kv".to_string());
+                    }
+
+                    if let std::collections::hash_map::Entry::Vacant(e) =
+                        capability_store.entry(c.name.clone())
```

Comment:
I would extract line 168 - 180 to a function, and reuse it in other capabilities.

Reply:
done ✅